### ISSUE TITLE
Auto-Calculation of a fare for a flight

### ISSUE THREAD

**Issue by lesmar54** [nabeelio/phpvms#1301]
**Is your feature request related to a problem? Please describe.**
If you are using the financials of PHPVMS you need to have a reasonable quality of fares for each flight and that takes a lot of research. Many have just set a basic fare for all flights I think or guessed a fare and the results in revenue can be a very large negative revenue or a very large positive revenue. Both of which do not sit well if you are trying to make the airline finances work with a model that looks good

**Describe the solution you'd like**
Given that airlines base their 'default' ticket prices on an agreed formula with IATA (Pat book) then it would come as no surprise that a mean average after research into 100's of thousands of flights, there was a basic fare that applied to all flights regardless of distance this was calculated at the time as being around $50. In addition to this the distance travelled and the airline average ticket cost was linear to the extent that the research found a per nautical mile cost of $0.11.
Now since PHPVMS has a fair simpler model in relation to fares I would like to suggest just like the distance between two 
airports can be auto-calculated it would be nice to do the same for a 'auto-calculated' fare for the flight.
Yes at present 
"right now you can set base fares and for flights/subfleets set them to a percent of the base fare"
However if you had the following factors to use for auto-calculation  for a flight then it can calculate the actual fare or percentage of a base fare.
So in a config section of the airline say under fares you have these -:
BASE_FARE i.e. $50
PER_MILE i.e $0.11
BUSINESS_FACTOR. i.e. 3. (business class is three times economy)
FIRST_FACTOR i.e. 5.  (first class is five times economy

Then when you create a flight you need a flag that has three values for auto fare pricing as follows
N - No auto pricing  default value which is basically saying use the system as it is now 
P - Auto pricing as a percentage of the base value 
V - Auto pricing to give a value for the fare
**Describe alternatives you've considered**
set base fares and for flights/subfleets set them to a percent of the base fare

**Additional context**
sample code on how it could be calculated , please note that any variables and names in the example are not in PHPVMS but just to show how the auto-calculation could work

//
//
// Auto fare calculation example
//
// This is calculated as follows -:
// $economybase = 135;  			this would be the value from the fares table for economy
// $businessbase = 500; 			this would be the value from the fares table for business
// $firstbase = 820; 				this would be the value from the fares table for business
// $basefare       i.e 50 			this would be from the fares config file 
// $permile        i.e 0.11			this would be from the fares config file
// $rounding       i.e 5			this would be from the fares config file
// $business       i.e 3			this would be from the fares config file
// $first          i.e 5			this would be from the fares config file
// $distance       i.e 1200			this would be from the flight either input or auto calculated 
//
// Economy Fare is Base Fare + (distance * per=mile) rounded to nearest rounding factor
// Business Fare is (Base Fare + (distance * per=mile)) * factor rounded to nearest rounding factor
// First Class Fare is (Base Fare + (distance * per=mile)) * factor rounded to nearest rounding factor
//
// function to round a number to nearest factor
//
	function roundUpToAny($n,$x) {
		return (round($n)%$x === 0) ? round($n) : round(($n+$x/2)/$x)*$x;
	}
//
// function to a percentage 
//
	function get_percentage($total, $number)
	{
		if ( $total > 0 ) {
			return round(($number * 100) / $total, 2);
		} else {
			return 0;
		}
	}
//
// dummy test results
//
	$economybase = 135;
	$businessbase = 500;
	$firstbase = 820;
	$distance = 1200;
	$basefare = 50;
	$permile  = 0.11;
	$business = 3;
	$first = 5;
	$rounding = 5;
// 
// this gives you the value amount for the fare
//
	$economy_fare = $basefare + ($distance * $permile);
	$economy_fare = roundUpToAny($economy_fare,$rounding);
//
//  this gives you the percentage amount 
//
	$economy_percent =  get_percentage($economybase,$economy_fare).'%'; 
	echo 'Economy Class fare for a '.$distance.'Nm flight is '.$economy_fare.' with percentage value of '.$economy_percent.PHP_EOL;
// 
// this gives you the value amount for the fare
//
	$business_fare = ($basefare + ($distance * $permile)) * $business;
	$business_fare = roundUpToAny($business_fare,$rounding);
//
//  this gives you the percentage amount 
//	
	$business_percent =  get_percentage($businessbase,$business_fare).'%';
	echo 'Business Class fare for a '.$distance.'Nm flight is '.$business_fare.' with percentage value of '.$business_percent.PHP_EOL;
// 
// this gives you the value amount for the fare
//
	$first_fare = ($basefare + ($distance * $permile)) * $first;
	$first_fare = roundUpToAny($first_fare,$rounding);
//
//  this gives you the percentage amount 
//
	$first_percent =  get_percentage($firstbase,$first_fare).'%';
	echo 'First Class fare for a '.$distance.'Nm flight is '.$first_fare.' with percentage value of '.$first_percent.PHP_EOL;

The output is as follows -:

Economy Class fare for a 1200Nm flight is 185 with percentage value of 137.04%
Business Class fare for a 1200Nm flight is 550 with percentage value of 110%
First Class fare for a 1200Nm flight is 910 with percentage value of 110.98%

**Comment by lesmar54** [comment]
I should say that in the maintenance part of PHPVMS you could process a global price re-calculation by changing the base fare that already exists in PHPVMS and that would work very well but you would or should also be able to amend the auto calculation config values as I defined above and auto re-calculate those flights that have an auto-calc flag of P and V

**Comment by nabeelio** [comment]
Thanks for adding this here. I will take a look through it in some more detail. I edited your comment just to fix the code blocks.

**Comment by FatihKoz** [comment]
If we can have an "auto calculation" option to enable/disable and "auto price/cost calculation for fares" happening when a flight starts (pirep prefile) or ends (pirep file) it will be more flexible.

With this, va admins will be controlling their base fare prices and costs via phpvms core fare system, defining their increase rate by distance and the enable/disable option from admin / settings. Which then be applied during pirep prefile/file stage and the auto adjusted prices will be used for finance of that pirep.

This will (and must) also cover the "cargo" fares used for that flight/pirep. 

I do not like the idea of "assigning fares" to flights and then changing their "price" by auto calculation.

And just a side note "distance travelled" should be the "great circle distance" between airports involved, not the "actual" air distance flown.

**Comment by FatihKoz** [comment]
Above is basically what I am doing at "pirep accepted" level ;) It is not tied to a setting, hard coded in my code and works well without any significant load during pirep acceptance.

**Comment by lesmar54** [comment]
My idea was a separate thing to the BASE FARE as input 

Now when you set up the flight you take the base fare and do several things to it 


So if you want a fare for that flight of 110 you would set the cost as 110%.   10% more than 100.
The auto calculation would take the flight distance from the flight as calculated or input 

And then work out the cost as a percentage in this case 

Economy Class fare for a 275.6Nm flight is 80 with percentage value of 80% 

It is only a guideline calculation that uses research values to help price flights.

Yes you could leave the fare at 100 I am not saying that every flight has to be priced individually  but when you want a fare that is a bit more realistic  than 100 for a 3000 mile flight . Unless you have searched the web sites for fares this would give you a simple and easy way 

So again by example the 3000 mile trip 
Economy Class fare for a 3000Nm flight is 380 with percentage value of 380% 

That was just adding a little bit of code to generate that all important value 

"Fares assigned to the current flight. These can be overridden, otherwise, the values used come from the subfleet of the aircraft that the flight is filed with. Only assign the fares you want to override. They can be set as a monetary amount, or a percentage. “

My suggested code works in the 'Only assign the fares you want to override’ context of the text when dealing with fares

**Comment by FatihKoz** [comment]
Understood, what your "auto pricing logic" offers is a value for flight fare. Either entered manually or automatically, but it is still defined at flight level and written to database.

What I offer for this idea is to make it fully automatic, calculated and applied automatically when a pirep gets accepted. 
('cause this is the stage/phase where actually this calculation is done in core phpvms)

Why, because defining flight level fares will make the management process harder when you have lots of flights. Of course it is applicable and nothing wrong with that if you want to spend time to manage them all.

Why should we calculate and write something to database and then never touch it again (until the time we want to change our price logic) ? Instead we can do that calculation on the fly when accepting a pirep.

"_Now when you set up the flight you take the base fare and do several things to it_"

No, I do not. I just create (or mostly) import the flights. I do not add fares to flights, I do have only base fares (which holds price and cost) and subfleet fares (which defines the capacities). [ And I am not the only one doing this :) ]

Your idea is good and will be a good addition to the core for sure. Hope it gets somehow implemented.

**Comment by lesmar54** [comment]
What I would be doing is just working out the fare as a percentage of the base fare in theory so if the base fare was 100 then my calculation would make the fare price in the flight as say 115% thus the fare would be 115 so if you changed the base fare to 200 then the fare would be 230.  I would not store the actual fare price but the %value  as per the guidelines 

“ Flight Fares# <https://docs.phpvms.net/#flight-fares>
If a fare is added to a flight, it will take precedence over the fares assigned to the subfleet.
When a fare is assigned to a flight, the value it can be given is either a monetary amount, or a multiplier. (e.g, 100, or 100%). The multiplier is applied from the subfleet amount, or the fare amount, if one isn't specified on the subfleet. By default, if there is no value, it is assumed at 100% of the amount.
If there are expected to be aircraft substitutions, you can add only the fares that may apply. For example, on a normal route, you may only have an Economy and First Class fare. But perhaps an aircraft with an additional fare, Premium Economy may be used. You can add Premium Economy to the flight, and set it to either a monetary amount, or to the multiplier, e.g, it might be set to 120%. When the PIREP is filed, for that flight, the seat prices for that flight, on that aircraft, will be set to 120% of the base premium economy price. "

So if you wanted a 5% increase in all fares then because for the miles calculation I use to generate a percentage amount then you would just add 5% to the PHPVMS base fare .   

That would work for now and still be within the logic of PHPVMS , however I agree with your potential idea of using the pirep to calculate the same value sounds good .  When the pirep is accepted of course the flight distance of the flight record should be used and not the pirep flight distance as if you flew the same flight twice with different sids and stars then the distance would be different and airlines don’t price the ticket on which way the wind is blowing.  Obviously auto-fare calculation at any stage is the lowest calculation to use and is an option and only overrides any other calculation if set to yes , the default must be to use the standard PHPVMS calculations as they have been defined and the pilot only selects to have a fare calculated.   

You would need perhaps to record the fact of auto-calculation in case the pilot gets say a negative revenue and asks why .  You may need to know that they selected auto-price rather than the PHPVMS calculations.

At least we are thinking that this could be a useful thing.

The good thing perhaps is to also develop a widget that shows you the fares in two formats The fare as calculated by PHPVMS that would be used in the PIREP reporting  and then this what I call industry average calculation  as a pointer .  If there is a big difference or one is lower than the other then the pilot gets some indication as to the revenue  from tickets.  Of course fares and just one item in calculating flight revenue as you have costs such as fuel and expenses and pilot pay and the widget would need somehow to calculate all the expenses potentially for a flight to give a. Revenue estimate.  Because you do not know how much fuel will be used for a flight the estimated revenue would be hard to forecast so just a fares compare would be simple and easy.

**Comment by FatihKoz** [comment]
A little example of what I tried to explain above... We need two steps main steps;

1. Build method/function to auto generate the price
```php
if (!function_exists('Dsp_AutoPrice')) {
  // Generate Automatic Fare Price Based on Distance
  // Return Numeric String
  function Dsp_AutoPrice($pirep, $fare, $base_price = 10, $per_nm = 0.11) {

    $gc_dist = Dsp_CalculateDistance($pirep->dpt_airport_id, $pirep->arr_airport_id);
    $distance = is_numeric($gc_dist) ? $gc_dist : $pirep->distance;

    if ($fare->type === FareType::CARGO) {
      $per_nm = round($per_nm / 110, 4);
      $base_price = 0.02;
    }

    if ($fare->code === 'J') { $multiplier = 3; }
    elseif ($fare->code === 'F') { $multiplier = 5; }
    else { $multiplier = 1; }

    return round($base_price + ($distance * $per_nm) * $multiplier, 2);
  }
}
```

2. Then use it when a Pirep's financial calculations are being done
```php
    public function payFaresForPirep($pirep): void
    {
        $fares = $this->getReconciledFaresForPirep($pirep);

        foreach ($fares as $fare) {

            if (setting('pireps.autocalculate_price', false)) {
                $fare->price = Dsp_AutoPrice($pirep, $fare, 'gc');
            }

            $credit = Money::createFromAmount($fare->count * $fare->price);
            $debit = Money::createFromAmount($fare->count * $fare->cost);
        }
    }
```
Base Ticket Price, Multipliers (for different fare types) can be tied to settings like the main function's usage... Also consider a solution for CARGO fares (like above) 'cause those Cargo fares have huge capacities (like 24000 kgs for a Freighter or 4000 kgs for a narrow body airliner), therefore using something like 55 as their cost would not work ;)

Long story short, I still do believe that this will be much practical instead of calculating prices and then writing them to database as flight fares (with a price tag or with some percentage values).

PS: We still may have a bug at csv import with modified fares, so people may not be able to export their flights with detailed fares and then import back.

**Comment by FatihKoz** [comment]
And just another addition to the main idea,

We are not selling tickets to someone and not displaying them somewhere. So we do not need to round up our fare prices to the nearest something :) However, we may need some nice decimals for those prices.

This is why, my above example have two digit rounding (and uses 4 digits for Cargo fare per nautical mile price)

**Comment by lesmar54** [comment]
Does the pirep pricing of fares/cargo as you explained override any PHPVMS values either from subfleet or flights ?.  If so then we are covered very well.  Because some VA may not want pilots to in effect calculate a fare based on an auto-calculation as maybe perhaps they have a ‘FOCUS’ airport for one month where they have added 5 to the PHPVMS fare price as a bonus for that month. 

So in my opinion I see it as follows 

PIREP has an option to auto-calculate a fare based on system parameters as described if the Pilot / VA allows the option to do so.
PIREP as a safety catch will auto-calculate a fare if the system does not generate a fare based on subfleet or flight.

If we have the variables stored at a system level i.e at a FARES table record for the variances i.e. base cost and per mile price and at the PHPVMS config level the control over auto-pricing then the flexibility is still there.

When setting up a flight and the VA wants to attach a fare to a flight then yes they can select the fare type and still like at present input an actual value but why not also allow the auto calculation of a percentage as an option.

That way we still keep the core design of the system intact as the only things being added are being used to create data for existing fields either on the main data tables or on submission of a pirep.

**Comment by lesmar54** [comment]
At present PHPVMSGEN does not calculate anything for CARGO flights as I only am setting up passenger only flights for a start ,  yes in subsequent enhancements this will be looked at in more detail and based upon the PIREP or PHPVMS calculation so as to conform to an agreed calculation.  
I do not mind changing anything in my project as it evolves and the more it can generate in a way that VA’s generally set up PHPVMS then the more happy new users will be with it .  
At the moment the system is driven by a small text file with just 15 values and I really just want to keep the number small as I do not want to have to test 40 or 50 conditions to see if what I build is correct. 
PHPVMSGEN does not replace the work a VA has to do to get and maintain a working Virtual Airline and it never will. The aim is just to provide the basic data fo 5 system files that once imported means that you have a simple ‘working’ airline where a lot of data input has been saved. Yes of course the names used by the project might not be to everyones liking but I cannot think for everyone.  One of the good uses of PHPVMSGEN would be to prototype a VA that you may be thinking of starting up without spending days and days spent in data input to see if it works.

**Comment by FatihKoz** [comment]
I am not talking about PHPVMSGEN or any other outside project/module here. I am only looking this matter as a feature for phpVMS v7.

Also for your question above; Yes, that code above overrides any fare price defined for a flight (or a subfleet) because it is running during a pirep's finance calculation process.

Pilots have no control over administrative things, they even do not see the fares (and they do not need to). It is only visible to admins at admin section. Maybe with some template tricks you can display a flight's fare (or pirep's transactions) somewhere in frontend.

And this is my last input to this topic, since it is in "Roadmap" as an "Enhancement" it is for the developer to think about it :) I shared my thoughts with an example, that is all ;)

**Comment by dougjuk** [comment]
I support this idea, I have implemented Dipos system via the special addon.  It would be nice if the core system allowed for the calculation to pass to an addon such as the Special addon with a switch or option.  Then the control of the fare can move away from individual flights and be more balanced.   Problem is right now it takes a core file alteration.
